x-pack/plugins/apm/server/routes/agent_explorer/get_agents_items.ts

```diff
@@ -0,0 +1,147 @@
+}
+
+export async function getAgentsItems({
+  environment,
+  agentLanguage,
+  serviceName,
+  kuery,
+  apmEventClient,
+  start,
+  end,
+  randomSampler,
+}: AggregationParams) {
+  const response = await apmEventClient.search('get_agent_details', {
+    apm: {
+      events: [ProcessorEvent.metric],
+    },
+    body: {
+      track_total_hits: false,
+      size: 0,
+      query: {
+        bool: {
+          filter: [
+            {
+              exists: {
+                field: AGENT_NAME,
+              },
+            },
+            {
+              exists: {
+                field: AGENT_VERSION,
+              },
+            },
+            ...rangeQuery(start, end),
+            ...environmentQuery(environment),
+            ...kqlQuery(kuery),
+            ...(serviceName ? termQuery(SERVICE_NAME, serviceName) : []),
+            ...(agentLanguage
+              ? termQuery(SERVICE_LANGUAGE_NAME, agentLanguage)
+              : []),
+          ],
+        },
+      },
+      aggs: {
+        sample: {
+          random_sampler: randomSampler,
+          aggs: {
+            services: {
+              terms: {
+                field: SERVICE_NAME,
+                size: MAX_NUMBER_OF_SERVICES,
+              },
+              aggs: {
+                instances: {
+                  cardinality: {
+                    field: SERVICE_NODE_NAME,
+                  },
+                },
+                serviceNodes: {
+                  terms: {
+                    field: SERVICE_NODE_NAME,
+                    size: MAX_NUMBER_OF_SERVICES,
+                  },
+                  aggs: {
+                    sample: {
+                      top_metrics: {
+                        metrics: [
+                          { field: AGENT_NAME } as const,
+                          { field: AGENT_VERSION } as const,
+                          { field: SERVICE_LANGUAGE_NAME } as const,
+                        ],
+                        sort: {
+                          '@timestamp': 'desc' as const,
+                        },
+                      },
+                    },
+                  },
+                },
+                environments: {
+                  terms: {
+                    field: SERVICE_ENVIRONMENT,
+                  },
+                },
+              },
+            },
+          },
+        },
+      },
+    },
+  });
+
+  return (
+    response.aggregations?.sample.services.buckets.map((bucket) => {
+      const agents = bucket.serviceNodes.buckets.map((serviceNode) => ({
+        name: serviceNode.sample.top[0].metrics[AGENT_NAME] as AgentName,
+        version: serviceNode.sample.top[0].metrics[AGENT_VERSION] as string,
+      }));
+
+      return {
+        serviceName: bucket.key as string,
+        environments: bucket.environments.buckets.map(
+          (env) => env.key as string
+        ),
+        agentName: map(agents, 'name')?.[0],
+        agentVersion: uniq(map(agents, 'version')),
```

Comment:
Wouldn't we have an inconsistency when querying the instances if we aggregate by `agent.version`?
Because for example we could have documents with many older versions of an agent but when we query the instances we are using the `top_metrics` so the instance would show the version used in the last document.

Reply:
https://user-images.githubusercontent.com/1313018/200026235-2e04344c-ff37-469a-b85e-59aafc921686.mov

Reply:
For the view that lists the instances, you'd need to break down on service.node.name and add the `missing` property to the terms agg. You can remove the random_sampler agg there - we're already looking at a subset of the data so it is less important there. Can you also mention this (that the random sampler doesn't allow `missing` for nested terms aggs) in the #es-analytics-geo channel to see if we should open up a feature request?

Reply:
I confirmed with #es-analytics-geo folks that `missing` is unsupported by design when using the sampler